### PR TITLE
Adding user defined metrics to kubelet `/stats/summary` API.

### PR DESCRIPTION
**Pull request by vishh**
This patch only includes the API.

Impl to follow in a separate PR.

### CODE REVIEW

File: pkg/kubelet/server/stats/types.go [hunk 1]
```diff
@@ -70,6 +72,8 @@ type PodStats struct {
 	// Stats pertaining to volume usage of filesystem resources.
 	// VolumeStats.UsedBytes is the number of bytes used by the Volume
 	VolumeStats []VolumeStats `json:"volume,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+	// User defined metrics that are exposed by containers in the pod. Typically, we expect only one container in the pod to be exposing custom metrics. In the event of multiple containers exposing metrics, they will be combined here.
```
Comment (timstclair) [inline pkg/kubelet/server/stats/types.go]: (nit) s/custom metrics/user defined metrics/

File: pkg/kubelet/server/stats/types.go [hunk 2]
```diff
@@ -147,3 +151,37 @@ type FsStats struct {
...
 }
+
+type UserDefinedMetricType string
+
+const (
+	// Instantaneous value. May increase or decrease.
+	MetricGauge UserDefinedMetricType = "gauge"
+
+	// A counter-like value that is only expected to increase.
+	MetricCumulative = "cumulative"
```
Comment (timstclair) [inline pkg/kubelet/server/stats/types.go]: (nit) You need to specify the type for each constant

File: pkg/kubelet/server/stats/types.go [hunk 3]
```diff
@@ -147,3 +151,37 @@ type FsStats struct {
...
+	// Rate over a time period.
+	MetricDelta = "delta"
+)
+
+type UserDefinedMetricValue struct {
+	// Time at which the metric was queried
+	Timestamp time.Time `json:"timestamp"`
+
+	// The value of the metric at this point.
+	Value *resource.Quantity `json:"value,omitempty"`
```
Comment (timstclair) [inline pkg/kubelet/server/stats/types.go]: Does it make sense to have a "MetricValue" without a value? I.e. should this be required?

File: pkg/kubelet/server/stats/types.go [hunk 4]
```diff
@@ -70,6 +72,8 @@ type PodStats struct {
 	// Stats pertaining to volume usage of filesystem resources.
 	// VolumeStats.UsedBytes is the number of bytes used by the Volume
 	VolumeStats []VolumeStats `json:"volume,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+	// User defined metrics that are exposed by containers in the pod. Typically, we expect only one container in the pod to be exposing custom metrics. In the event of multiple containers exposing metrics, they will be combined here.
+	UserDefinedMetrics []UserDefinedMetric `json:"userDefinedMetrics,omitmepty" patchStrategy:"merge" patchMergeKey:"name"`
```
Comment (timstclair) [inline pkg/kubelet/server/stats/types.go]: hmm, this structure is a little funny... now "sample" is a top-level construct (the whole summary is a single time-stamped sample), it seems a little funny to have multiple UserDefinedMetric samples with different time stamps nested within the structure. How will we decide which `UserDefinedMetricValue`s (samples) to include with the summary?

**Comment by vishh** [comment]
@timstclair: I updated this PR and rebased it on top of #19932. PTAL

**Comment by timstclair** [comment]
Finished reviewing.

**Comment by vishh** [comment]
Updated the patch based on comments.

**Comment by vishh** [comment]
Once this PR is finalized, I'll close this and merge it into #19932.

**Comment by timstclair** [comment]
LGTM

**Comment by vishh** [comment]
Closing this in favor of https://github.com/pwittrock/kubernetes/pull/1. This PR will be folded into #19932 to work around the submit queue delays :P

**Pull request closed without merging**